File: cpp/samples/core/imonumber.cpp
/**
 * @file imonumber.cpp
 * @brief Demonstrates usage of vista-sdk ImoNumber API
 * @details This sample shows how to create, validate, and parse IMO numbers
 *          with checksum verification per IMO Resolution A.600(15)
 */

#include <dnv/vista/sdk/core/ImoNumber.h>

#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

int main()
{
    using namespace dnv::vista::sdk;

    std::cout << "=== vista-sdk ImoNumber Sample ===\n\n";

    {
        std::cout << "1. ImoNumber: Creating from integer\n";
        std::cout << "--------------------------------------\n";

        // Create valid IMO numbers
        auto imo1 = ImoNumber{ 9074729 };
        auto imo2 = ImoNumber{ 9785811 };
        auto imo3 = ImoNumber{ 1234567 };

        std::cout << "Created IMO numbers:\n";
        std::cout << "  " << imo1.toString() << "\n";
        std::cout << "  " << imo2.toString() << "\n";
        std::cout << "  " << imo3.toString() << "\n";

        // Try invalid number (will throw)
        std::cout << "\nAttempting to create invalid IMO number (1234507):\n";
        try
        {
            [[maybe_unused]] auto invalid = ImoNumber{ 1234507 };
            std::cout << "  ERROR: Invalid number accepted (unexpected)\n";
        }
        catch (const std::invalid_argument& e)
        {
            std::cout << "  Correctly rejected: " << e.what() << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "2. ImoNumber: Creating from string\n";
        std::cout << "-------------------------------------\n";

        // With "IMO" prefix
        auto imo1 = ImoNumber{ "IMO9074729" };
        std::cout << "From 'IMO9074729': " << imo1.toString() << "\n";

        // Without "IMO" prefix
        auto imo2 = ImoNumber{ "9785811" };
        std::cout << "From '9785811'   : " << imo2.toString() << "\n";

        // Both create equivalent objects
        std::cout << std::boolalpha;
        std::cout << "Are they equal? " << (imo1 == ImoNumber{ "IMO9074729" }) << "\n";
        std::cout << "Are they equal? " << (imo2 == ImoNumber{ "IMO9785811" }) << "\n";

        // Try invalid string (will throw)
        std::cout << "\nAttempting to create from invalid string ('IM9074729'):\n";
        try
        {
            [[maybe_unused]] auto invalid = ImoNumber{ "IM9074729" };
            std::cout << "  ERROR: Invalid string accepted (unexpected)\n";
        }
        catch (const std::invalid_argument& e)
        {
            std::cout << "  Correctly rejected: " << e.what() << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "3. ImoNumber::isValid(): Validating integers\n";
        std::cout << "-----------------------------------------------\n";

        std::vector<int> testNumbers = {
            9074729,  // Valid
            9785811,  // Valid
            1234567,  // Valid
            1234507,  // Invalid checksum
            123456,   // Too short
            12345678, // Too long
            0,        // Invalid
            -1        // Invalid
        };

        std::cout << std::boolalpha;
        std::cout << "Validation results:\n";
        for (int num : testNumbers)
        {
            bool valid = ImoNumber::isValid(num);
            std::cout << "  " << std::setw(10) << num << ": " << valid << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "4. ImoNumber::fromString(): Safe parsing without exceptions\n";
        std::cout << "-------------------------------------------------------------\n";

        std::vector<std::string> testStrings = {
            "9074729",   "IMO9785811", "1234567", // Valid
            "1234507",                            // Invalid checksum
            "IM9074729",                          // Typo in prefix
            "",                                   // Empty
            "abc"                                 // Not a number
        };

        std::cout << "Parsing results:\n";
        for (const auto& str : testStrings)
        {
            auto result = ImoNumber::fromString(str);

            std::cout << "  '" << str << "'" << std::string(10 - str.length() + 1, ' ') << "-> ";

            if (result.has_value())
            {
                std::cout << result->toString() << "\n";
            }
            else
            {
                std::cout << "Invalid\n";
            }
        }

        std::cout << "\n";
    }

    {
        std::cout << "5. ImoNumber::fromString(): Optional usage patterns\n";
        std::cout << "----------------------------------------------------\n";

        std::string input = "IMO9074729";

        // Pattern 1: Direct if with optional
        if (auto imo = ImoNumber::fromString(input))
        {
            std::cout << "Parsed successfully: " << imo->toString() << "\n";
        }
        else
        {
            std::cout << "Parse failed\n";
        }

        // Pattern 2: has_value() check
        auto result = ImoNumber::fromString("9785811");
        if (result.has_value())
        {
            std::cout << "Parsed successfully: " << result.value().toString() << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "6. ImoNumber: Comparison operators\n";
        std::cout << "-------------------------------------\n";

        auto imo1 = ImoNumber{ 9074729 };
        auto imo2 = ImoNumber{ "IMO9074729" };
        auto imo3 = ImoNumber{ 9785811 };

        std::cout << std::boolalpha;
        std::cout << "Equality:\n";
        std::cout << "  9074729 == IMO9074729? " << (imo1 == imo2) << "\n";
        std::cout << "  9074729 == 9785811   ? " << (imo1 == imo3) << "\n";

        std::cout << "\nInequality:\n";
        std::cout << "  9074729 != 9785811   ? " << (imo1 != imo3) << "\n";
        std::cout << "  9074729 != IMO9074729? " << (imo1 != imo2) << "\n";

        std::cout << "\n";
    }

    {
        std::cout << "7. ImoNumber: toString() formatting\n";
        std::cout << "--------------------------------------\n";

        auto imo = ImoNumber{ 9074729 };

        std::cout << "String representation:\n";
        std::cout << "  toString(): " << imo.toString() << "\n";
        std::cout << "  Format    : Always includes \"IMO\" prefix\n";
        std::cout << "  Format    : Always 7 digits after \"IMO\"\n";

        // Different input formats, same output
        std::cout << "\nConsistent output:\n";
        std::cout << "  From int 9074729 : " << ImoNumber{ 9074729 }.toString() << "\n";
        std::cout << "  From '9074729'   : " << ImoNumber{ "9074729" }.toString() << "\n";
        std::cout << "  From 'IMO9074729': " << ImoNumber{ "IMO9074729" }.toString() << "\n";

        std::cout << "\n";
    }

    {
        std::cout << "8. ImoNumber: Real-world examples\n";
        std::cout << "------------------------------------\n";

        // Real IMO numbers
        struct Ship
        {
            std::string name;
            int imoNumber;
        };

        std::vector<Ship> ships = { { "Queen Mary 2", 9241061 },
                                    { "Peter Faber", 8027781 },
                                    { "Oasis of the Seas", 9383936 },
                                    { "Symphony of the Seas", 9744001 } };

        std::cout << "Famous ships:\n";
        for (const auto& ship : ships)
        {
            if (ImoNumber::isValid(ship.imoNumber))
            {
                auto imo = ImoNumber{ ship.imoNumber };
                std::cout << "  " << std::setw(20) << std::left << ship.name;
                std::cout << ": " << imo.toString() << "\n";
            }
        }

        std::cout << "\n";
    }

    {
        std::cout << "9. ImoNumber: Error handling patterns\n";
        std::cout << "-----------------------------------------\n";

        std::vector<std::string> inputs = { "9074729", "1234507" }; // Valid, Invalid

        for (const auto& userInput : inputs)
        {
            std::cout << "\nTesting with '" << userInput << "':\n";

            std::cout << "  Pattern 1: Constructor with try-catch (throws on error):\n";
            try
            {
                auto imo = ImoNumber{ userInput };
                std::cout << "    Success: " << imo.toString() << "\n";
            }
            catch (const std::invalid_argument& e)
            {
                std::cout << "    Error: " << e.what() << "\n";
            }

            std::cout << "  Pattern 2: fromString() with optional (no exceptions):\n";
            if (auto imo = ImoNumber::fromString(userInput))
            {
                std::cout << "    Success: " << imo->toString() << "\n";
            }
            else
            {
                std::cout << "    Error: Invalid IMO number format\n";
            }
        }

        std::cout << "\n";
    }

    return 0;
}

File: bindings/rust/vista-sdk/examples/core/imo_number.rs
use vista_sdk::core::imo_number::ImoNumber;

fn main() {
    println!("=== vista-sdk ImoNumber Sample ===\n");

    {
        println!("1. ImoNumber: Creating from integer");
        println!("----------------------------------------");

        // Create valid IMO numbers
        let imo1 = ImoNumber::create(9074729).expect("9074729 should be a valid IMO number");
        let imo2 = ImoNumber::create(9785811).expect("9785811 should be a valid IMO number");
        let imo3 = ImoNumber::create(1234567).expect("1234567 should be a valid IMO number");

        println!("Created IMO numbers:");
        println!("  {}", imo1);
        println!("  {}", imo2);
        println!("  {}", imo3);

        // Try invalid number (returns Err)
        println!("\nAttempting to create invalid IMO number (1234507):");
        match ImoNumber::create(1234507) {
            Ok(_) => println!("  ERROR: Invalid number accepted (unexpected)"),
            Err(e) => println!("  Correctly rejected: {}", e),
        }

        println!();
    }

    {
        println!("2. ImoNumber: Creating from string");
        println!("---------------------------------------");

        // With "IMO" prefix
        let imo1 = ImoNumber::create_from_str("IMO9074729")
            .expect("IMO9074729 should be a valid IMO number");
        println!("From 'IMO9074729': {}", imo1);

        // Without "IMO" prefix
        let imo2 =
            ImoNumber::create_from_str("9785811").expect("9785811 should be a valid IMO number");
        println!("From '9785811'   : {}", imo2);

        // Both create equivalent objects
        println!(
            "Are they equal? {}",
            imo1 == ImoNumber::create_from_str("IMO9074729").unwrap()
        );
        println!(
            "Are they equal? {}",
            imo2 == ImoNumber::create_from_str("IMO9785811").unwrap()
        );

        // Try invalid string (returns Err)
        println!("\nAttempting to create from invalid string ('IM9074729'):");
        match ImoNumber::create_from_str("IM9074729") {
            Ok(_) => println!("  ERROR: Invalid string accepted (unexpected)"),
            Err(e) => println!("  Correctly rejected: {}", e),
        }

        println!();
    }

    {
        println!("3. ImoNumber::is_valid(): Validating integers");
        println!("---------------------------------------------------");

        let test_numbers = [
            9074729,  // Valid
            9785811,  // Valid
            1234567,  // Valid
            1234507,  // Invalid checksum
            123456,   // Too short
            12345678, // Too long
            0,        // Invalid
            -1,       // Invalid
        ];

        println!("Validation results:");
        for num in test_numbers {
            let valid = ImoNumber::is_valid(num);
            println!("  {:>10}: {}", num, valid);
        }

        println!();
    }

    {
        println!("4. ImoNumber::from_str(): Safe parsing without exceptions");
        println!("-------------------------------------------------------------");

        let test_strings = [
            "9074729",
            "IMO9785811",
            "1234567",   // Valid
            "1234507",   // Invalid checksum
            "IM9074729", // Typo in prefix
            "",          // Empty
            "abc",       // Not a number
        ];

        println!("Parsing results:");
        for s in test_strings {
            let padding = 10usize.saturating_sub(s.len()) + 1;
            print!("  '{}'{}-> ", s, " ".repeat(padding));

            match ImoNumber::from_str(s) {
                Some(imo) => println!("{}", imo),
                None => println!("Invalid"),
            }
        }

        println!();
    }

    {
        println!("5. ImoNumber::from_str(): Optional usage patterns");
        println!("----------------------------------------------------");

        let input = "IMO9074729";

        // Pattern 1: Direct if with optional
        if let Some(imo) = ImoNumber::from_str(input) {
            println!("Parsed successfully: {}", imo);
        } else {
            println!("Parse failed");
        }

        // Pattern 2: is_some() check
        let result = ImoNumber::from_str("9785811");
        if result.is_some() {
            println!("Parsed successfully: {}", result.unwrap());
        }

        println!();
    }

    {
        println!("6. ImoNumber: Comparison operators");
        println!("-------------------------------------");

        let imo1 = ImoNumber::create(9074729).expect("9074729 should be a valid IMO number");
        let imo2 = ImoNumber::create_from_str("IMO9074729")
            .expect("IMO9074729 should be a valid IMO number");
        let imo3 = ImoNumber::create(9785811).expect("9785811 should be a valid IMO number");

        println!("Equality:");
        println!("  9074729 == IMO9074729? {}", imo1 == imo2);
        println!("  9074729 == 9785811   ? {}", imo1 == imo3);

        println!("\nInequality:");
        println!("  9074729 != 9785811   ? {}", imo1 != imo3);
        println!("  9074729 != IMO9074729? {}", imo1 != imo2);

        println!();
    }

    {
        println!("7. ImoNumber: toString() formatting");
        println!("----------------------------------------");

        let imo = ImoNumber::create(9074729).expect("9074729 should be a valid IMO number");

        println!("String representation:");
        println!("  to_string(): {}", imo);
        println!("  Format     : Always includes \"IMO\" prefix");
        println!("  Format     : Always 7 digits after \"IMO\"");

        // Different input formats, same output
        println!("\nConsistent output:");
        println!(
            "  From int 9074729 : {}",
            ImoNumber::create(9074729).unwrap()
        );
        println!(
            "  From '9074729'   : {}",
            ImoNumber::create_from_str("9074729").unwrap()
        );
        println!(
            "  From 'IMO9074729': {}",
            ImoNumber::create_from_str("IMO9074729").unwrap()
        );

        println!();
    }

    {
        println!("8. ImoNumber: Real-world examples");
        println!("--------------------------------------");

        struct Ship {
            name: &'static str,
            imo_number: i32,
        }

        let ships = [
            Ship {
                name: "Queen Mary 2",
                imo_number: 9241061,
            },
            Ship {
                name: "Peter Faber",
                imo_number: 8027781,
            },
            Ship {
                name: "Oasis of the Seas",
                imo_number: 9383936,
            },
            Ship {
                name: "Symphony of the Seas",
                imo_number: 9744001,
            },
        ];

        println!("Famous ships:");
        for ship in ships {
            if ImoNumber::is_valid(ship.imo_number) {
                let imo = ImoNumber::create(ship.imo_number).expect("validated above");
                println!("  {:<20}: {}", ship.name, imo);
            }
        }

        println!();
    }

    {
        println!("9. ImoNumber: Error handling patterns");
        println!("-----------------------------------------");

        let inputs = ["9074729", "1234507"]; // Valid, Invalid

        for user_input in inputs {
            println!("\nTesting with '{}':", user_input);

            println!("  Pattern 1: create_from_str() (returns Result on error):");
            match ImoNumber::create_from_str(user_input) {
                Ok(imo) => println!("    Success: {}", imo),
                Err(e) => println!("    Error: {}", e),
            }

            println!("  Pattern 2: from_str() with Option (no exceptions):");
            match ImoNumber::from_str(user_input) {
                Some(imo) => println!("    Success: {}", imo),
                None => println!("    Error: Invalid IMO number format"),
            }
        }

        println!();
    }
}
